### ISSUE TITLE
Problem when setData is called during manipulation

### ISSUE THREAD

**Issue by felixhayashi** [almende/vis#568]
Hi guys,

I have a problem here :( 

**Problem Context**

Sometimes a set of graphs (simultaniosly displayed) in the tiddlywiki is being triggered to fully refresh as some articles (represented as nodes) where added or a bunch or relationships suddenly ceased to exist.

This may happen automatically e.g. due to an ajax request that loads new articles or because the user added a bunch of articles (by writing or drag'n'drop) while he/she was in "manipulation" mode.

I cannot selectively refresh the graphs saying "remove this edge and this node and add these nodes" but instead rely on `network.setData` to update the graph with the new data.

**Actual Problem**

Calling setData while the manipulation mode is active creates problems. See demo here
http://jsbin.com/mehowu/2/

**Possible solutions**

My first choice would be to programmatically break out of manipulation mode before calling `network.setData()`. I can simulate the "back button" click via

    var backButton = this.parentDomNode.getElementsByClassName("network-manipulationUI back")[0];
    backButton.click();

This works fine -- almost. In case of the *edit edge mode* mode, I get an error (but despite the error it then continues to work as expected)

    Uncaught TypeError: Cannot read property 'controlNodes' of undefined

As said, giving me a chance to exit the edit mode via a function would solve the problem.
Another solution would be that in such a situation (as described above) vis automatically aborts the manipulation as the underlying datastructure changed.

Maybe you have a better idea how to handle the case when setData is called during manipulation? - I cannot prevent it from happening as otherwise the user deals with outdated data. The problem generally exists for all systems that persist, load and remove data in bulk (from an external store) into the graph regardless what the user is doing at that time.

Regards Felix

**Comment by AlexDM0** [comment]
Hi Felix,

Thats an annoying one... In any case, I'll take a look. The solution you have there does work for the editedges mode.. Try not using the minified version. We don't always keep it up to date on develop.

Regards

**Comment by AlexDM0** [comment]
I've pushed a fix to develop for this. A cleanup call is done in setData if the user has dataManipulation enabled. (similar to pressing back or escape)

**Comment by felixhayashi** [comment]
Oh this works really sweet plus it is a good generic solution and also exactly what I had in mind when I wrote

> Another solution would be that in such a situation (as described above) vis automatically aborts the manipulation as the underlying datastructure changed.

@AlexDM0, you solve these problems with incredible speed. great job :+1:

**Comment by AlexDM0** [comment]
Hi Felix,

You're welcome :). Just as a general idea, wouldn't it be better if you use the datasets for your updating instead of setData?

Regards,

Alex

**Comment by felixhayashi** [comment]
Hi @AlexDM0,

**TL;DR**

I did that for a long time and it worked great! But it became hard to maintain. Now it's easier to erase all data and rebuild everything from scratch.

**Long story**

when I started the project, I maintained a very fine grained decision mechanism:

After initialization of the graphs, I checked for article and relationship changes and always used the `vis.DataSet` to reflect the changes via `add`, `update` and `remove` (both for edges and nodes). While it worked great, it was already difficult since I allow the user to assemble graphs via a filter mechanism and have to keep track of all situations that impact the graph. Adding up to that comes the problem, that I am only roughly informed about changes in the wiki. For example I cannot distinguish between a title change and a new insert. So I had to differentiate between many events that all impact the graph differently.

Two months ago, I then added the possibility to display neighbourhoods that do not directly match a filter but are connected to nodes that do match. This introduced another level of complexity and made it even harder to keep this "fine grained decision mechanism": Now I had to check which edges were added to existing nodes that are not neighbours themselves that connect nodes that are not included in the set described by the filter. Also once a node is removed I had to check whether this affects the set of neighbours. I ended up with a lot of code in order to keep my fine grained update mechanism alive. Hard to maintain. It scared me.

After doing some performance tests and comparing different use cases of the plugin, I finally realised that it is a much cleaner solution to simply rebuild the graph (completely) when an article or a connection has changed (doesn't happen too frequently, so it's fine). Vis.js handles a rebuild incredibly fast (even when 10 graphs rebuild simultaniously, each containing ~50 nodes). This way I can calculate the graph and the neighbours based on the current system's state and do not have to maintain a sophisticated update mechanism.

-Felix

**Comment by AlexDM0** [comment]
Hi Felix,

Thank you for the explanation! You have an interesting project going on! 

Now there is something I was wondering. Since you have vis running and thus know what data is in it, and you have new data, why not cross reference at that stage and insert update and delete accordingly? That way you can keep your current infrastructure and the user just continues what he's doing. There might be an obvious reason why this is not possible but from your story I didn't immediately see it. 

I agree that keeping track of everything is a pain, but with your current setup there comes a point where you have old data and new data. 

We could also argue this may be a nice addition to the current dataset. A method that compares current data with new data and performs the required actions to make sure the current data becomes the new one. 

I'd be interested in hearing your thoughts on this. 

Regards,

Alex

**Comment by felixhayashi** [comment]
Hi @AlexDM0,

sorry, this post became rather long. But I feel better giving you an elaborated answer than risking that you think I accidentally made some bad design choices. I want to mention beforehand that your suggestion to delegate the choice, how to deal with the data, to vis is truly a great idea (see bottom of my post).

> You have an interesting project going on! 

Thanks :) It's also a part of my master-thesis (design science), which is, broadly speaking, about how to publicly distribute knowledge at low setup-costs.

> Now there is something I was wondering. Since you have vis running and thus know what data is in it, and you have new data, why not cross reference at that stage and insert update and delete accordingly?

Yes, this is close to what I did at the beginning. The problem is, for various reasons (see further below "Updating Style 2") I *cannot* simply forward the changes like so:

* article added -> nodes.add()
* article deleted -> nodes.remove()
* article updated -> nodes.update()

> That way you can keep your current infrastructure and the user just continues what he's doing.

I completely agree. Keeping the current infrastructure would be the best. I mean, this is what the vis.DataSet is made for: to automatically reflect changes, instead of always removing everything and rebuilding from scratch. Until recently, I tried really hard to maintain what you call the "current infrastructure" (see further below).

> I agree that keeping track of everything is a pain, but with your current setup there comes a point where you have old data and new data. 

Nope, don't worry, it's all fine :) I react to all changes and they are immediately reflected it's just that I refresh all nodes or edges *completely* when the system state changed instead of doing a *selective refresh*. So it is rather a performance question.

> There might be an obvious reason why this is not possible but from your story I didn't immediately see it. 

Sorry, maybe my last post failed to explain the issue. Unfortunately yes, there is a reason, yet, it is not easily explained. I actually spend many days thinking about it before I changed my "architecture" last month.

But I will give it a try.

Just a quick note on the vocab:
------------------------------

*graph filter* = A filter on which basis a graph decides which articles to display as nodes. 
*matches* = Nodes that match a graph's filter and are therefore included and displayed.
*neighbours* = nodes that do not match the graph's filter themselves, however, are directly ("level 1 neighbours") connected to nodes that match the filter.

Updating Style 1: "Rebuild from scratch" 
------------------------------

If I perform a complete refresh (via `setData()`) I can simply run the following (pseudo) code:

    NODES = Load all nodes from store that match a specified filter
    
    if(show neighbours)
      EDGES = Load all edges from store that that have at least one endpoint in the NODES set
      NEIGHBOURS = all nodes that represent an endpoint in EDGES and are not contained in NODES
    else
      EDGES = Load all edges from store that have both endpoints contained in the NODES set
      
That is all. Easy to maintain, understandable.

At runtime I only need to check

* Did an article change (e.g. because the user clicked "save article changes") that is represented in the graph or matches the filter? **->** rebuild graph
* Did an article connection change anywhere in the system? **->** rebuild graph

Again, very easy and it's actually very fast (as said in the previous post, even when 10 graphs rebuild simultaneously, each containing ~50 nodes).

Of course one might argue a complete rebuild is an overkill. Yes, I completely agree and I am highly aware of it.

But in a standard scenario (aka "real life") it doesn't matter:

* articles do not update all the time
* not many graphs are simultaneously displayed in the wiki
* Graphs used in TiddlyWiki are usually rather smaller graphs with ~20 nodes.
* computers are fast, vis is fast

But what matters in real live is that code needs to be understandable, maintainable, easy.

So what would be the alternative? Below is a simplified version of what I did before I opted for the mechanism described above. 

Updating Style 2: Dynamic updating
------------------------------

Please, if your eyes start bleeding because of the heavily nested lists, just quickly skim through the various points, to get an idea of the complexity of the decision tree.

The following function is called on every graph instance by the tiddlywiki system when any article or connection changed at any place in the system.
The changed_articles parameter is a list of articles that only contains information on whether an article has been changed (no further differentiation into add or update) or removed.

**Graph.prototype.refresh(changed_relations, changed_articles) {**

**for each** changed article / changed relation:

  * If article changed (updated or added; not removed from system)
    * Does the article match the filter so it is accepted as graph-node?
      * Yes, it matches
        * Does the article already exist in the current nodes-set?
          * If not, add the article to the vis.DataSet for graph nodes
            * In order to display neighbours, load all edges ("tempEdgesSet") that are connected to this node<br/> **At this point I need to loop over all the edges in the system, (actively maintaining a lookup table is an overkill)**
              * Load all nodes that represent an endpoint in "tempEdgesSet" and are not with both endpoints connected to the node we are looking at. <br />
                Also no nodes are allowed to be loaded that are already contained in the existing overall nodes-set.
                  * If they are not contained, they receive a "neighbour" group property value
          * If yes, it means the article has been updated
            * Is the node that represents the article currently displayed as neighbour?
              * If yes, a neighbour turned into a match. Change the group property value from "neighbour" to "match"
              * Otherwise it already exists as "match" in the nodes-set; simply update
      * No, article doesn't match the filter
        * Scan all nodes in the current nodes-set for whether the article is represented as node in the graph
          * If it exists as a neighbour
            * Check the current nodes-set if it is still a valid neighbour of some node
              * If yes, update it
              * If no, remove it
          * If it exists as a part of the "match" group, it was once a match but not anymore.
            * get all connected edges
              * if edge leads to a neighbour, remove the neighbour and the edge
              * If the edge leads to a match, only remove the edge
  * If an article has been removed from the system, has it been displayed in the graph before?
    * If it has been displayed (and thus is contained in the current nodes-set)
      * Does it exist as match or as neighbour
        * If it exists as neighbour, simply remove neighbour and edge
        * Otherwise
          * get all connected edges
            * if edge leads to a neighbour, remove the neighbour and the edge
            * If the edge leads to a match, only remove the edge
    * If it hasn't been included in the graph before
      * Irrelevant, do nothing
  * If an edge has changed (means updated or added; not removed)
    * If the edge existed before (= update)
      * Simply update the edge, doesn't really change anything but the layout or the edge label.
    * If the edge is newly added
      * Is the edge connected to a node in the nodes dataset?
        * Is it connected to a neighbour?
          * Do nothing as we only display level 1 neighbourhoods
        * Is it connected to a "match" node in the nodes-set?
          * get the node from the store that represents the other endpoint 
            * If this node is not contained in the graph yet, display it as neighbour
  * In case an edge is removed
      * Is it connected to any article that is represented in the nodes-set?
        * If yes
          * if a connected node is a neighbour, also remove the neighbour-node from the nodes-set
          * if a connected node is not a neighbour, only remove the edge

I created this decision tree roughly from my former code and from memory so I apologize if it doesn't cover all cases (or contains mistakes). Also what I didn't include in this decision tree are some special cases that would need to be further considered (e.g. certain nodes are always displayed by force). Also imagine if I displayed not only direct neighbours but transitive neighbours. It would become even more complicated.

---

> We could also argue this may be a nice addition to the current dataset. A method that compares current data with new data and performs the required actions to make sure the current data becomes the new one.
> I'd be interested in hearing your thoughts on this. 

This of course would be great, honestly! I mean in this case, I could simply pass the data and the decision to vis! While I would still use the "Updating Style 1" since it is much more maintainable, I do not need (and want) to erase all the current data. Vis could decide which data to remove, replace or add based on the provided ids or on the missing ids.

I am definetly a fan of that. Something like `nodes.refresh([ ... ])` in addition to add, remove and update.

Regards

Felix

**Comment by AlexDM0** [comment]
Hi Felix,

Thanks for the exploitation :). I'm still not exactly sure this can't be done easier. Let me try to explain. 

You start a network with edges and nodes on page load. 

Semi pseudo code:

```
var nodesdata = dataset with nodes
var edgesdata = dataset with edges

var network = network
```

You probably have an AJAX call to get new data periodically, let's call it getNewData. Let's assume the data you get is a JSON with nodes and edges fields. Each of those containing JSON values. 

So there comes a point where:

``` 
var newData = getNewData();
```

You then have both 
```
newData.nodes
newData.edges

var nodesjson = nodesData.get( as json )
var edgesjson = edgesData.get( as json )
```

We can then brute force check easily. This is cheap since you have very few nodes. 

```
Var nodesUpdateQueue = []
Var nodesDeleteQueue = []
For node in newData.nodes
  If node.id exists in nodesjson
    nodesUpdateQueue.push(node)
  else
    nodesDeleteQueue.push(node)
```

Now edges can be tricky since they may not have an id. So: 

```

Var edgesUpdateQueue = []
Var edgesDeleteQueue = []
For edge in newData.edges
  If  edge exists with same to and from in edgesjson AND handled is false
    edgesUpdateQueue.push(edge)
    Edgesjson[edge we found that makes this true].handled = false
  else
    EdgesDeleteQueue.push(edge)
```

Finally update the datasets with the 4 arrays we constructed. 

Why can't this be done? It is only dependent on the data you now put in setData and the existing one. It can't fail and it is cheap to do computationally, especially compared to the recalculation of the physics. 

The main reason for this is that the nodes do not initialise in the same way so the user would see jumps in the network where you do a setData. 

What am I missing that makes this not feasible?

Regards,

Alex

**Comment by AlexDM0** [comment]
Haha I mean explanation instead of exploitation. Great typing on phones :)

**Comment by felixhayashi** [comment]
> Thanks for the exploitation :)
> Haha I mean explanation instead of exploitation. Great typing on phones :)

This typo also made me laugh quite a bit :)

> Why can't this be done? It is only dependent on the data you now put in setData and the existing one. It can't fail and it is cheap to do computationally, especially compared to the recalculation of the physics. 

You know what? You are right.  -.- 

I completely oversaw this solution. I mean just because I don't do a fine grained "dynamic updating" doesn't mean I can't iterate over the node-changes (or edge-changes) in the end and simply use the existing datasets to add, remove and update data. And as you said, this is just a cheap operation with linear complexity. I will certainly implement it.

Thanks for unflinchingly pointing at my mistake, Alex! ;)

-Felix

**Comment by AlexDM0** [comment]
;) you're welcome. If you did that initially we would never have found the bug with the manipulation :)

**Comment by AlexDM0** [comment]
By the way, update also adds if the node does not exist. 

Be careful with the edges though. Ids are used to match so do get the Id from the old list and append it to the new if you want to update. Otherwise you'll end up with a LOT of double edges.

**Comment by felixhayashi** [comment]
> ;) you're welcome. If you did that initially we would never have found the bug with the manipulation :)

true :)

> By the way, update also adds if the node does not exist. 

Yap, I know, so I can be cheap and just use update instead of differentiating here between update and add as it is not needed in my case.

> Be careful with the edges though. Ids are used to match so do get the Id from the old list and append it to the new if you want to update. Otherwise you'll end up with a LOT of double edges. 

I am aware of that but thanks for pointing that out.

Have a great Weekend.

-Felix

**Comment by AlexDM0** [comment]
I'll add the merge function to the dataset feature requests. Cheers!

**Comment by felixhayashi** [comment]
Nice, thanks!

I already implemented it on my own, but having a merge function is better!

I know, you will handle this in a different way but just for the record, I already implemented the following and it works as we wanted it to work, keeping the existing infrastructure alive and critically reducing the changes that are handed over to the network :)

    /**
     * using an existing dataset to reflect the changes between
     * two node sets.
     * 
     * @param {Hashmap<id, Node>} lt1 - Lookup table that contains the
     *     *new* set of nodes.
     * @param {Hashmap<id, Node>} lt1 - lookup table that holds the
     *     *old* set of nodes.
     * @param {vis.DataSet} [ds] - The dataset to be updated
     */
    utils.refresh = function(lt1, lt2, ds) {
      
      if(!ds) {
        return new vis.DataSet(utils.convert(lt1, "array"));
      }
          
      var changes = {
        updated: utils.convert(lt1, "array"),
        removed: []
      };
      
      for(var id in lt2) {
        if(!lt1[id]) changes.removed.push(id);
      }
      
      ds.remove(changes.removed); // remove before update!
      ds.update(changes.updated);
      
      return ds;
      
    };

-Felix
